src/metabase/driver/mongo.clj

```diff
@@ -42,17 +41,21 @@
     #".*"                               ; default
     message))
 
-(defn- process-query-in-context [_ qp]
+(defn- process-query-in-context [qp]
   (fn [query]
-    (with-mongo-connection [^DB conn, (:database query)]
-      (qp query))))
+    (let [{:keys [database], :as query} (update query :database (fn [database]
+                                                                  (if (integer? database)
+                                                                    (Database database)
```

Comment:
Asking the driver to lookup the database on it's own is proliferating the problem of broken driver encapsulation.  It's far more consistent and sustainable to have a QP middleware function resolve the database object before the query makes it to the driver.

Since every query we run requires a lookup of the Database details its being executed against, by centralizing this in the QP middleware we remove redundant code and promote better encapsulation of the drivers.

Feel free to do this in a separate PR, but now between the generic-sql, druid, mongo, and crate drivers this is becoming a more widespread anti-pattern.